web/editor_undo_bar.js

```diff
@@ -0,0 +1,95 @@
+/* Copyright 2024 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+class EditorUndoBar {
+  #container;
+
+  #controller = null;
+
+  isOpen = false;
+
+  #message;
+
+  #undoButton;
+
+  static #l10nMessages = Object.freeze({
+    highlight: "pdfjs-editor-undo-bar-message-highlight",
+    freetext: "pdfjs-editor-undo-bar-message-freetext",
+    stamp: "pdfjs-editor-undo-bar-message-stamp",
+    ink: "pdfjs-editor-undo-bar-message-ink",
+    _multiple: "pdfjs-editor-undo-bar-message-multiple",
+  });
+
+  constructor({ container, message, undoButton, closeButton }, eventBus) {
+    this.#container = container;
+    this.#message = message;
+    this.#undoButton = undoButton;
+
+    // Caveat: we have to pick between registering these everytime the bar is
+    // shown and not having the ability to cleanup using AbortController.
+    const boundHide = this.hide.bind(this);
+    closeButton.addEventListener("click", boundHide);
+    eventBus._on("beforeprint", boundHide);
+    eventBus._on("download", boundHide);
+    eventBus._on("annotationstart", boundHide);
+    eventBus._on("secondarytoolbaraction", boundHide);
```

Comment:
While this suggestion would be slight abuse of notation, do we really need a new event just for this or could we use the already existing one instead?
```suggestion
    eventBus._on("reporttelemetry", evt => {
      if (evt?.details?.type === "buttons") {
        boundHide();
      }
    });
```

Reply:
We originally considered this, but yeah it felt a bit too much abusing an event for something that's not meant to. Ideally an event called `"reporttelemetry"` should not have UI effects other than... reporting telemetry 😅.

Happy to change it if you feel strongly though.

Reply:
Why not just add a click listener on the window (capture) and check if the target is under the toast bar ?
It's something we do in order to hide the secondary toolbar:
https://github.com/mozilla/pdf.js/blob/master/web/app.js#L2750-L2765

Reply:
I meant the latter.
The main problem of having this list of event is that we've to "maintain" it, I mean if we add something in the future, we'll have to remember to update it.